src/lib/shmem/src/raw_syscall.rs

```diff
@@ -0,0 +1,222 @@
+//! This module provides a simple interface to make system calls without relying on libc or nix,
+//! which are not no-std. The goal of this module is to be as no-std as possible to be compatible
+//! with Shadow's shim preload library.
+//!
+//! Public functions for system calls are named according to their standard names found in man(2).
+//! The public function accept have parameters with types that are natural for Rust. Most of these
+//! public functions are accompanied by a *_impl() function that has a signature which more closely
+//! matches the x86_64 ABI:
+//! <https://blog.rchapman.org/posts/Linux_System_Call_Table_for_x86_64/>
+
+use std::arch::asm;
+
+unsafe fn interpret_str_to_char_ptr(s: &str) -> *const char {
+    unsafe { core::mem::transmute::<*const u8, *const char>(s.as_bytes().as_ptr()) }
+}
+
+fn interpret_i32_as_u64(x: i32) -> u64 {
+    u64::from_ne_bytes((x as i64).to_ne_bytes())
+}
+
+fn interpret_u64_as_i32(x: u64) -> i32 {
+    let x_bytes = x.to_ne_bytes();
+    if cfg!(target_endian = "big") {
+        i32::from_ne_bytes(x_bytes[4..8].try_into().unwrap())
+    } else {
+        i32::from_ne_bytes(x_bytes[0..4].try_into().unwrap())
+    }
+}
+
+fn convert_i32_rv_to_rv_errno(rv: i32) -> Result<i32, i32> {
+    if rv < -1 {
```

Comment:
This isn't right for, e.g. `mmap`. It should be something like `if (-4095..=-1).contains(rv)`

https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/x86_64/sysdep.h;h=24d8b8ec20a55824a4806f8821ecba2622d0fe8e;hb=HEAD#l41

Reply:
fixed